addons/sale_timesheet/models/account.py

```diff
@@ -11,78 +11,61 @@ class AccountAnalyticLine(models.Model):
             values['so_line'] = task.sale_line_id.id or values.get('so_line', False)
-        for line in self:
-            values.update(line._get_timesheet_values(values))
-            values.update(line._get_timesheet_billing_values(values))
-            super(AccountAnalyticLine, line).write(values)
-
-        # Update delivered quantity on SO lines which are not linked to the analytic lines anymore
-        so_lines -= self.mapped('so_line')
-        if so_lines:
-            so_lines.with_context(force_so_lines=so_lines).sudo()._compute_analytic()
-        return True
+        return values
 
-    def _get_timesheet_values(self, values):
-        result = {}
-        values = values if values is not None else {}
-        if values.get('project_id') or self.project_id:
-            if values.get('amount'):
-                return {}
-            unit_amount = values.get('unit_amount', 0.0) or self.unit_amount
-
-            # find the employee
-            employee_id = values.get('employee_id') or self.employee_id.id or (self.env.user.employee_ids and self.env.user.employee_ids[0].id or False)
-            employee = self.env['hr.employee'].browse(employee_id) if employee_id else self.env['hr.employee']
-            if not employee:
-                user_id = values.get('user_id') or self.user_id.id or self._default_user()
-                employee = self.env['hr.employee'].search([('user_id', '=', user_id)], limit=1)
-
-            # find the analytic account
-            account_id = values.get('account_id') or self.account_id.id
-            if not account_id and values.get('project_id'):
-                project = self.env['project.project'].browse(values['project_id'])
-                account_id = project.analytic_account_id.id
-            analytic_account = self.env['account.analytic.account'].browse(account_id) if account_id else self.env['account.analytic.account']
-
-            # convert employee cost into timesheet (analytic account) currency
-            uom = employee.company_id.project_time_mode_id
-            cost = employee.timesheet_cost or 0.0
-            amount = -unit_amount * cost
-            amount_converted = employee.currency_id.compute(amount, analytic_account.currency_id)
+    @api.multi
+    def _timesheet_postprocess(self, values):
+        # (re)compute the amount (depending on unit_amount, employee_id for the cost, and account_id for currency)
+        if any([field_name in values for field_name in ['unit_amount', 'employee_id', 'account_id']]):
+            for timesheet in self:
+                uom = timesheet.employee_id.company_id.project_time_mode_id
+                cost = timesheet.employee_id.timesheet_cost or 0.0
+                amount = -timesheet.unit_amount * cost
+                amount_converted = timesheet.employee_id.currency_id.compute(amount, timesheet.account_id.currency_id)
+                timesheet.sudo().write({
```

Comment:
Another suggestion / remark: isn't it possible to use update in order to work in cache so that _timesheet_compute_theorical_revenue make the final write containing all values ?